controllers/helmrelease_controller.go

```diff
@@ -252,189 +82,135 @@ func (r *HelmReleaseReconciler) SetupWithManager(mgr ctrl.Manager, opts HelmRele
+		}
+		r.recordReadiness(hr, false)
+		return ctrl.Result{}, err
 	}
 
-	// Determine last release revision.
-	rel, observeLastReleaseErr := run.ObserveLastRelease(hr)
-	if observeLastReleaseErr != nil {
-		err = fmt.Errorf("failed to get last release revision: %w", observeLastReleaseErr)
-		return v2.HelmReleaseNotReady(hr, v2.GetLastReleaseFailedReason, "failed to get last release revision"), err
+	// Examine if our object is under deletion
+	if !hr.ObjectMeta.DeletionTimestamp.IsZero() {
+		result, err := r.reconcileDelete(ctx, run, hr)
+		if err != nil {
+			return result, err
+		}
+		r.recordReadiness(hr, true)
+		if err := r.Client.Update(ctx, hr); err != nil {
+			return result, err
+		}
+		return result, retErr
 	}
 
-	// Register the current release attempt.
-	revision := chart.Metadata.Version
-	releaseRevision := util.ReleaseRevision(rel)
-	valuesChecksum := util.ValuesChecksum(values)
-	hr, hasNewState := v2.HelmReleaseAttempted(hr, revision, releaseRevision, valuesChecksum)
-	if hasNewState {
-		hr = v2.HelmReleaseProgressing(hr)
-		if updateStatusErr := r.Status().Update(ctx, &hr); updateStatusErr != nil {
-			log.Error(updateStatusErr, "unable to update status after state update")
-			return hr, updateStatusErr
+	// Record reconciliation duration
+	if r.MetricsRecorder != nil {
+		objRef, err := reference.GetReference(r.Scheme, hr)
+		if err != nil {
+			return ctrl.Result{}, err
 		}
-		// Record progressing status
-		r.recordReadiness(hr, false)
+		defer r.MetricsRecorder.RecordDuration(*objRef, reconcileStart)
 	}
 
-	// Check status of any previous release attempt.
-	released := meta.GetCondition(hr.Status.Conditions, v2.ReleasedCondition)
-	if released != nil {
-		switch released.Status {
-		// Succeed if the previous release attempt succeeded.
-		case corev1.ConditionTrue:
-			return v2.HelmReleaseReady(hr), nil
-		case corev1.ConditionFalse:
-			// Fail if the previous release attempt remediation failed.
-			remediated := meta.GetCondition(hr.Status.Conditions, v2.RemediatedCondition)
-			if remediated != nil && remediated.Status == corev1.ConditionFalse {
-				err = fmt.Errorf("previous release attempt remediation failed")
-				return v2.HelmReleaseNotReady(hr, remediated.Reason, remediated.Message), err
-			}
+	// Always attempt to patch the status after each reconciliation
+	defer func() {
+		// Record the value of the reconciliation request
+		if v, ok := meta.ReconcileAnnotationValue(hr.GetAnnotations()); ok {
+			hr.Status.LastHandledReconcileAt = v
 		}
-
-		// Fail if install retries are exhausted.
-		if hr.Spec.GetInstall().GetRemediation().RetriesExhausted(hr) {
-			err = fmt.Errorf("install retries exhausted")
-			return v2.HelmReleaseNotReady(hr, released.Reason, released.Message), err
+		// Record observed generation
+		hr.Status.ObservedGeneration = hr.Generation
+		if retErr != nil {
+			v2.HelmReleaseNotReady(hr, meta.ReconciliationFailedReason, err.Error())
```

Comment:
Are we bubbling up the reasons/messages (e.g. install/upgrade/test failures) to the Ready condition yet like we do today?

Reply:
This is bubbling up the error(s) returned by the `reconcile*` methods to the `Ready` message.

Reply:
Ok, so just not bubbling up the `Reason`s yet. That would be very useful to keep as well if we can. For example anything that uses the `kstatus` library which consumes only the `Ready` condition, will then be able to surface the underlying `Reason` which provides a quick "at a glance" diagnosis for the end user, and some level of machine readability of the status.